reddit/baseplate.py
# Add default Prometheus metrics for Thrift server

This original [PR](https://github.com/reddit/baseplate.py/pull/647) was getting too big, so this PR splits out one part. There will be approx. 4 PRs split out in total:
* 1st PR (this PR) adds default Prometheus metrics for Thrift server
* 2nd PR will do the same for Thrift Client
* 3rd PR will add default Prometheus metrics for HTTP server
* 4th PR will do the same for HTTP client

Background:
We are adding support to export Prometheus metrics from baseplate.py services.  As of baseplate.py [v2.3.0](https://github.com/reddit/baseplate.py/releases/tag/v2.3.0), `baseplate-serve` also runs a server that exports Prometheus metrics on endpoint `/metrics` on port 6060. 

This PR adds the default Prometheus metrics for Thrift servers. The default Thrift metrics to be exported are defined in the [baseplate.spec](https://github.snooguts.net/reddit/baseplate.spec/blob/master/component-apis/prom-metrics.md#thrift).

Testing:
I make a test thrift app from baseplate-cookiecutter and tested out this new Prometheus code. Here are the steps I used to test:
1) create thrift service with baseplate-cookiecutter (make sure baseplate v2.3.0 or greater) so that prom metrics are exported on port 6060 by default.
2) make sure `prometheus-client==0.12.0` is in the requirements.txt file
3) set env var `PROMETHEUS_MULTIPROC_DIR` to a directory where metrics from multiprocess python can be written  (this is already included in prod [dockerfile](https://sourcegraph.build.ue1.snooguts.net/github.snooguts.net/reddit/docker-baseplate-py/-/blob/2/Dockerfile?L25:5))
4) either set `metrics.tagging` or `metrics.namespace` in the service's ini config file.
5) build thrift with `make thrift` then run `baseplate-serve example.ini`
6) use the test-client to hit the `is_healthy` endpoint
7) check the prometheus metrics that were created:
```
$ curl localhost:6060/metrics
# HELP thrift_server_active_requests Multiprocess metric
# TYPE thrift_server_active_requests gauge
thrift_server_active_requests{pid="23318",thrift_method="is_healthy"} 0.0
# HELP thrift_server_latency_seconds Multiprocess metric
# TYPE thrift_server_latency_seconds histogram
thrift_server_latency_seconds_sum{thrift_method="is_healthy",thrift_success="true"} 0.041624628
thrift_server_latency_seconds_bucket{le="0.0001",thrift_method="is_healthy",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.00025",thrift_method="is_healthy",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.000625",thrift_method="is_healthy",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.0015625",thrift_method="is_healthy",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.00390625",thrift_method="is_healthy",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.009765625",thrift_method="is_healthy",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.0244140625",thrift_method="is_healthy",thrift_success="true"} 0.0
thrift_server_latency_seconds_bucket{le="0.06103515625",thrift_method="is_healthy",thrift_success="true"} 1.0
thrift_server_latency_seconds_bucket{le="0.152587890625",thrift_method="is_healthy",thrift_success="true"} 1.0
thrift_server_latency_seconds_bucket{le="0.3814697265625",thrift_method="is_healthy",thrift_success="true"} 1.0
thrift_server_latency_seconds_bucket{le="0.95367431640625",thrift_method="is_healthy",thrift_success="true"} 1.0
thrift_server_latency_seconds_bucket{le="2.384185791015625",thrift_method="is_healthy",thrift_success="true"} 1.0
thrift_server_latency_seconds_bucket{le="5.9604644775390625",thrift_method="is_healthy",thrift_success="true"} 1.0
thrift_server_latency_seconds_bucket{le="14.901161193847656",thrift_method="is_healthy",thrift_success="true"} 1.0
thrift_server_latency_seconds_bucket{le="+Inf",thrift_method="is_healthy",thrift_success="true"} 1.0
thrift_server_latency_seconds_count{thrift_method="is_healthy",thrift_success="true"} 1.0
# HELP thrift_server_requests_total Multiprocess metric
# TYPE thrift_server_requests_total counter
thrift_server_requests_total{thrift_baseplate_status="",thrift_baseplate_status_code="",thrift_exception_type="",thrift_method="is_healthy",thrift_success="true"} 1.0
```

It appears the `HELP` text is clobbered in multiprocess mode, [issue](https://github.com/prometheus/client_python/issues/211).
